### ISSUE TITLE
Confused with the overridestate design decision

### ISSUE THREAD

**Issue by MisterFOnGitHub** [Unity-Technologies/PostProcessing#456]
I'm confused with this part of the design. OverrideState is said to be per volume, but I didn't find a way to modify the default value of an effect in the profile asset itself. What's the point of overriding ? For instance if you ad a grain effect, intensity is 0 by default, so you NEED to override it (would you use a zeor intensity grain filter ??)

In practice, you end up overriding a lot of settings in the profile, and even with one global volume with a profile that is never modified via script, the PostProcessLayer.PreCull() may end up taking more time than the actual rendering of the post fx... (I profiled it and a lot of time is spent overriding the settings in the OverrideSettings method, especially for spline parameters, the whole precull taking up to 0.6 or 0.7ms on a Core i7).

Did I miss something ? How can a setting be overriden per volume if it is actually modifying the value in the profile asset ?

**Comment by MisterFOnGitHub** [comment]
Looking at it a bit closely, it seems that it might be fixable by processing BuildCommandBuffers in another thread, in parallel to SceneCulling. PostProcessLayer.OnPreRender could then wait for BuildCommandBuffers to be done.
@chman, do you see any reason why putting it in a thread wouldn't work (I haven't looked into every bit of code) ? If not, I'll take the time to prototype that and see if it works well.

**Comment by Chman** [comment]
It's a global override stack, there's no "default state" per profile. Think of it as a big set of deck of cards, with the top-most card being the one with the highest priority and each card representing one parameter. The default state would be the table itself, or the very first card in each deck (it created automatically).

> What's the point of overriding ? For instance if you ad a grain effect, intensity is 0 by default, so you NEED to override it (would you use a zeor intensity grain filter ??)

Yes, you need a zero intensity grain filter (0 intensity isn't being rendered anyway so it costs nothing on the GPU). If you don't, how do you return to a "null" state if at some point the effect has to be turned off? You'd have to manually set intensity to 0 in all of your volumes.

Basic use-case: you have a global volume without any grain. Somewhere in a level, you want to apply a grain effect but only to a specific set of rooms, so you create a local volume with Grain, set a priority and a blending radius so it smoothly interpolates the grain when you get close to it. If you didn't have a default state of 0 you wouldn't get the effect you'd expect ("turn on the grain when I get to the volume, turn it off when I leave"). You would have to explicitly add Grain to your global volume and manually override intensity to 0 to make sure it interpolates back nicely to 0.

The only effect that can't really have a "null state" is Depth of Field.

In the end you're still able to override the default state by creating a global volume with a very low priority and adding/overriding everything. But it's not as easy as simply having a volume "marked as being the default with no need to override". There are too many UX & workflow implications to list them all here.

Granted, documentation is currently lacking but it's being written as we speak and will be there for the 2018.1 release. In the meantime I highly recommend the [Quickstart guide](https://github.com/Unity-Technologies/PostProcessing/wiki/%28v2%29-Quickstart) (which gives some idea about why things are the way they are) and [this big comment](https://github.com/Unity-Technologies/PostProcessing/blob/v2/PostProcessing/Runtime/PostProcessVolume.cs#L6...L51) (rapid overview of the volume stuff).

> I profiled it and a lot of time is spent overriding the settings in the OverrideSettings method

Did you profile in the editor? Because in practice I see much lower numbers in builds, the profiler adds quite a bit of overhead in the editor (test on an i7 5960X and PS4 - single core performance isn't the best there).

> do you see any reason why putting it in a thread wouldn't work (I haven't looked into every bit of code) ?

Yep, it'll probably break hard as most of Unity's user-facing API isn't thread-safe at all and we use it extensively when we build command buffers. Feel free to try though, but I wouldn't hold my breath :/

**Comment by MisterFOnGitHub** [comment]
Ok thanks for the extensive reponse, it's really appreciated.

Other than that, don't get me wrong, I think the new post processing stack is great. I'm just a bit picky on the details to help make it as good as possible. Keep up the good work !

(oh one more little thing, I've seen the bloom bleeding to white on the sides in VR/single pass, not sure if that's a known bug but i thought I'd mention it...) It doesn't happen consistently so o couldn't figure out what exactly was causing it though.

**Comment by Chman** [comment]
Hmm the white bleeding in VR might be related to #392? If so then yes that would make it a known bug :)

I'll close this issue as it's more a Q&A type of thing, but don't hesitate to reply here if you experiment with threads etc.
